### ISSUE TITLE
Several crafting recipes depend on moreores

### ISSUE THREAD

**Issue by FaceDeer** [minetest-mods/technic#480]
There's a number of crafting recipes in this mod that depend on inputs from the moreores mod, but there aren't `minetest.get_modpath("moreores")` guards around them or alternative crafting recipes in the event that moreores is not installed (rendering these items uncraftable).

namely: [blue_energy_crystal](https://github.com/minetest-mods/technic/blob/8ebc1de29c2c5105ed4348fd8e93d3787dd106d9/technic/crafts.lua#L73), [red_energy_crystal](https://github.com/minetest-mods/technic/blob/8ebc1de29c2c5105ed4348fd8e93d3787dd106d9/technic/crafts.lua#L82), [mithril and silver dust](https://github.com/minetest-mods/technic/blob/8ebc1de29c2c5105ed4348fd8e93d3787dd106d9/technic/machines/register/grinder_recipes.lua#L108), [prospector](https://github.com/minetest-mods/technic/blob/8ebc1de29c2c5105ed4348fd8e93d3787dd106d9/technic/tools/prospector.lua#L121), and [sonic_screwdriver](https://github.com/minetest-mods/technic/blob/8ebc1de29c2c5105ed4348fd8e93d3787dd106d9/technic/tools/sonic_screwdriver.lua#L90).

**Comment by Xiong** [comment]
I think it's a valid issue but I'd rephrase it and just require a hard dep on moreores. 

Until reading this issue and going to check, I had always assumed this was already the case.

**Comment by FaceDeer** [comment]
Technic already has a large number of hard dependencies that I find offputting, surely there's some other material already in Technic that could be used as a substitute for mithril when that mod isn't installed? My modding philosophy is that hard dependencies should only be for requirements that _cannot_ be worked around with optional alternatives.

**Comment by Xiong** [comment]
I agree with  FaceDeer that deps are often bad. Code reuse is good but random dep networks become nightmares. 

However technic makes little sense without a large number of special materials. One way or another, their availability (however rare) must be ensured.

**Comment by FaceDeer** [comment]
Perhaps technic could use some refactoring to try compartmentalizing it a bit more, which could help with managing dependencies and making it easier to pick and choose which bits of technic to install. It seems a bit odd that if I want to add nuclear reactors using excruciatingly-detailed uranium formulations (I count 306 recipes created by the technic mod solely dedicated to mixing different kinds of uranium dust together) I also have to have magical metal from Tolkein's Middle Earth installed.

I've been working on a mod lately [(simplecrafting_lib)](https://github.com/minetest-mods/simplecrafting_lib/tree/toward_1.0) that incidentally allows me to do large-scale analysis of a Minetest game's crafting recipes, which is how I noticed the unsatisfied mithril dependency (I just threw a bunch of mods together into a test world to give myself a lot of diverse recipes to work with). Out of curiousity, I just now tried installing the _minimum_ hard dependency requirements for technic (just `default`, `basic_materials`, `pipeworks`, `mesecons`, and `mesecons_mvps`) and found a few more recipes that depend on things that are only registered by optional dependencies:

- `technic:solar_panel` requires `mesecons_materials:glue`
- `technic:lv_compressor' requires `mesecons:piston`
- `technic:tool_workshop` requires `mesecons_pistons:piston_sticky_off`
- `technic:sonic_screwdriver` requires `mesecons_materials:fiber` (as well as the previously mentioned mithril ingot)

Then I tried each of the technic modules separately. When just `concrete` and its hard dependencies are installed:

- `technic:blast_resistant_concrete` requires `technic:composite_plate`

When just `extranodes` and its hard dependencies are installed:

- `technic:insulator_clip` requires `technic:raw_latex`
- `technic:insulator_clip_fencepost` requires `technic:raw_latex`

There were also some pipeworks recipes that depended on moreores and on mesecon modules that aren't listed as hard dependencies, so this isn't just a technic problem. And my code currenly only just checks whether the required materials _exist_, not whether they're themselves craftable, so there could be other things that aren't tripping my error checks but that a player would be unable to craft in an actual game due to broken precursor recipes. Would it be helpful if I did some work trying to graph out these recipe networks and maybe write up a more detailed issue report?

**Comment by FaceDeer** [comment]
Incremental improvements are still improvements, I'll never be unhappy with them. :)

**Comment by Xiong** [comment]
I'd be very interested to see a decently hi-res PNG of this. I have not found an Android app that will render it.

**Comment by FaceDeer** [comment]
Here it is rendered as a PNG:
![technic](https://user-images.githubusercontent.com/19756022/52180304-e2cfd500-27a1-11e9-8a31-7a60aff53d96.png)

Probably not as useful as the graphml version, in yEd you can click around and see what connects to what in a "Neighborhood" window and search for specific nodes in a list. But it gives an overview, at least. I'll fiddle around a bit with the layout of the main group of recipes (the ones that don't have uranium in them) and see if I can find one that's a bit less visually tangled.

**Comment by FaceDeer** [comment]
I've been manually disentangling the non-uranium section of the graph and it's starting to come together a little:

![technic minus uranium](https://user-images.githubusercontent.com/19756022/52181413-b66e8580-27ae-11e9-9bab-7f8e2a64c8d0.png)

[technic minus uranium.zip](https://github.com/minetest-mods/technic/files/2825908/technic.minus.uranium.zip)

I take it this mod's recipes have grown somewhat organically over time? I'm seeing some patterns and natural groupings, but also a few oddities. The chainsaw is the only item that requires an on/off button, for example, which seems like something a lot of other tools could use too. The lava can requires zinc, which I find a bit odd since zinc is a low-melting-point metal, and I assume it's because the lava can was added at the same time as the water can which also requires it. Glue is only used to make solar panels, but glass is not - glass is only used to make cnc. Grinders have the peculiar requirement of default:desert_stone, I'm not sure what the motivation behind that is. There's an "island" of chest recipes made from various types of metals that don't seem connected to the rest of technic either thematically or as dependencies, and I see they've already been extracted into their own sub-mod, are they part of this modpack due to historical happenstance?

This is straying a bit far afield from just the mithril dependency, feel free to kick me into a different issue or to rename this one. I have to admit that I don't actually *use* technic, so the significance of some of these patterns is probably lost on me. But that might also give me a useful outside perspective, so if you like I could go on poking around and asking questions about some of these.

**Comment by Xiong** [comment]
I appreciate the PNG; it supports the overall contention that the graph is complex. 

But perhaps what's most needed is less detail. So long as a craft remains within a single mod, there's no issue in getting it to run. Intermod deps are critical. Unclear to me when (A deps on B) requires B loads before A but I smell trouble with mutual or circular dependency. Which... is likely only a matter of time.
